### DISCUSSION TITLE
File Stream Data Presist Issue

### DISCUSSION THREAD

**Discussion by AkthemRehab** [apache/streampipes#2284] [Q&A] [answered]
Hello Streampipes,
I create a new file stream adapter as follows:

<img width="1800" alt="Screenshot 2023-12-06 at 5 40 13 PM" src="https://github.com/apache/streampipes/assets/22823265/6b56ae8a-0f19-4b65-aa88-08c28d71ce01">

The file I upload is attached: 
[test.csv](https://github.com/apache/streampipes/files/13588130/test.csv)

I then add the timestamp and choose to persist events. When I create a new data view in the data explorer as follows:

<img width="1800" alt="Screenshot 2023-12-06 at 5 44 32 PM" src="https://github.com/apache/streampipes/assets/22823265/73aeb786-cfdf-4e41-a0ba-682019f90ef8">

I find the number of raws much less than expected. Doing some investigation I find that number of raws matching the # Events in "Settings - Data Lake".

However, checking the Data Lake Metrics, I find 20,480 consumed as expected. Refer below:

<img width="1800" alt="Screenshot 2023-12-06 at 5 46 39 PM" src="https://github.com/apache/streampipes/assets/22823265/7e8c6a39-fde2-48a1-9057-610ce1818aeb">

It only worked properly once, but other than that it keeps returning differing number of rows with inconsistent value each time. Could you please help me get this right? Or at least understand such behavior?

Thank you,
Akthem

**Comment by tenthe** [comment, marked as answer]
Hi @AkthemRehab,
thanks for your question.

The problem you are facing is due to the fact that the File Stream Adapter is primarily designed for testing purposes. It operates by taking a CSV file and replaying its contents. When there's no "original" timestamp present in the data, the adapter generates a new timestamp for each row upon reading. Consequently, this results in multiple events sharing the same timestamp. When these events are stored in the time series database, entries with identical timestamps overwrite each other. As a result, your data lake sink will receive the correct number of events, but not all of them will be stored in the database. The timestamp is used as an index in time series databases. If two events arrive with the same time stamp, they are overwritten.

To address this, consider providing an 'original' timestamp to your CSV data. By doing so, StreamPipes will utilize this timestamp, and you should observe the desired outcome.

Does this explanation clarify the issue for you? 
If you have any additional questions, feel free to ask.

Cheers,
Philipp

**Reply by AkthemRehab** [reply]
Hi Philipp @tenthe ,
Thanks for your informative reply. It helps a lot with my Streampipes first baby steps!

Now I understand why I see the behavior. I am wondering if the _file stream_ has an option to give a unique timestamp for each line though? The file is for a vibration reading and should have 'one' original timestamp. Usually the name of the file is the timestamp (eg. the file I shared is originally named: 2004.02.12.21.22.39) and the readings within the file are taken within the span of 1 second.

Let me elaborate more on my objective . Ultimately, I want to ingest similar files to Streampipes, forward the **files** as an http post to an api, and then collect the response from the api and present it in the dashboard/data explorer. I would also benefit a lot from the ability to storing the file in Streampipes and showing the time wave form, FFT, and other signal processing values in the Streampipes dashboard and data explorer.

Later on I will like to ingest the files directly from a sensor as an MQTT message, HTTP post, or perhaps other protocol. 

I hope I am not overwhelming you by widening the scope of my original question and my newbie level of questions. 

Much appreciate the help, informative, and educational answers/thoughts.

Best,
Akthem

**Reply by AkthemRehab** [reply]
Hey @tenthe ,
I marked your previous reply as answer acknowledging my follow up is perhaps another set of questions. Please let me know if a new thread would be preferable in this regard?

Best
Akthem

**Reply by tenthe** [reply]
Hi, thanks for marking the question as answered. 
Yes, I think it makes sense to open new threads for the follow-up questions.
I see three topics:
- Dealing with files in the File Stream Adapter that don't have a timestamp on every line. (e.g. adding an option to create timestamps at a fixed interval)
- Handling files (or arrays of events) as data source in the streaming environments
- List of needed algorithms for working with event streams

Please correct me if I have misunderstood something. You are also welcome to add any other topics you are interested in. 
My suggestion would be to discuss each topic in a separate thread to avoid mixing things up. 
What do you think about this?

**Reply by AkthemRehab** [reply]
Hi Philipp, You got it all right. I am wondering though what you mean by the third topic "List of needed algorithms for working with event streams"? Does it have to do with the http post containing file as pay load?

I will create two threads for the first two topics assuming they have not been discussed/addressed before. And once I get a better understanding of what you mean by the third, I will proceed with creating its thread too.

**Reply by tenthe** [reply]
Thank you for opening the discussion. I will answer you directly there.
On the third topic: I thought you are missing any algorithms to transform the data as processing elements within a pipeline.
If that is the case just start a discussion. If not we can ignore this topic for now.

**Reply by AkthemRehab** [reply]
Hi Philipp, I think I am more worried about the ability to HTTP POST with a file as the payload (multipart). Similarly ability to pub/sub file payloads in the MQTT nodes. Not sure if that requires algorithms that are currently missing?

**Comment by tenthe** [comment]
No, that's something else. Let's discuss it in the other discussion.
Thanks
